core/src/main/java/org/elasticsearch/search/internal/ContextIndexSearcher.java

```diff
@@ -152,7 +152,7 @@ public void search(Query query, Collector collector) throws IOException {
             // throws Lucene.EarlyTerminationException when given count is reached
             collector = Lucene.wrapCountBasedEarlyTerminatingCollector(collector, searchContext.terminateAfter());
         }
-        if (currentState == Stage.MAIN_QUERY) {
+        if (currentState == Stage.MAIN_QUERY && (query == searchContext.query() || query == searchContext.parsedQuery().query())) {
```

Comment:
@jpountz This is doable. The way it is going to look:
1) Add a `es_search(query, collector)`  method to ContextIndexSearch and move the logic where it is now to this new method. (minus the state management)
2) On places where  do search we invoke this new method. (QueryPhase, AggregationPhase and Lucene#count(...), ScanContext etc.)
3) In places where we invoke the IndexSearcher methods without an collector (to get top docs / search after etc.) We construct the collector ourselves and pass it down to the new method. This is a bit of a downside, but maybe a better tradeoff? (this only seems to be the case in QueryPhase).

I kind of like this change, but it is way bigger than this entire PR now. So can it be added as a follow up PR? If not then I like to get the first commit of this PR in if that is okay (since that doesn't require the change in ContextIndexSearcher).